src/query_plan/fetch_dependency_graph.rs

```diff
@@ -321,6 +640,18 @@ impl FetchSelectionSet {
             selection_set,
         })
     }
+
+    fn add_at_path(
+        &mut self,
+        path_in_node: &OpPath,
+        selection_set: Option<&Arc<NormalizedSelectionSet>>,
+    ) -> Result<(), FederationError> {
+        Arc::make_mut(&mut self.selection_set).add_at_path(path_in_node, selection_set);
+        // TODO: when calling this multiple times, maybe only re-compute conditions at the end?
+        // Or make it lazily-initialized and computed on demand?
+        self.conditions = self.selection_set.conditions()?;
```

Comment:
Here we re-compute conditions for the entire selection set for every call to `FetchSelectionSet::add_at_path`, which could have O(n²) cost. Should conditions be lazily initialized (with `add_to_path` only emptying the cache) to avoid this?